File: apps/ios/Tests/CrossyUITests/AmbientClockTests.swift
import Foundation
import XCTest

@testable import CrossyUI

// The ambient timer (ID-2, apps/ios/DESIGN.md §9): 0:00 quietly before the first
// fill, ticking from `firstFillAt` (root DESIGN.md D15: the origin is wire data,
// zero updates), frozen at completion. The clock is injected, so every case pins an
// exact string.

final class AmbientClockTests: XCTestCase {
    private let origin = "2026-07-10T19:02:11Z"

    private func date(_ iso: String) -> Date {
        AmbientClock.parse(iso)!
    }

    func test_parse_acceptsPlainAndFractionalISO8601_protocol3() {
        let plain = AmbientClock.parse("2026-07-10T19:02:11Z")
        let fractional = AmbientClock.parse("2026-07-10T19:02:11.123Z")
        XCTAssertNotNil(plain)
        XCTAssertNotNil(fractional)
        // The fractional stamp is 123 ms after the plain one.
        XCTAssertEqual(fractional!.timeIntervalSince(plain!), 0.123, accuracy: 0.0001)
        XCTAssertNil(AmbientClock.parse("not a time"))
    }

    func test_beforeFirstFill_readsZeroQuietly_id2() {
        XCTAssertEqual(
            AmbientClock.display(firstFillAt: nil, completedAt: nil, now: date(origin)),
            "0:00")
    }

    func test_ticksFromFirstFillAt_d15() {
        let now = date(origin).addingTimeInterval(754)
        XCTAssertEqual(
            AmbientClock.display(firstFillAt: origin, completedAt: nil, now: now),
            "12:34")
    }

    func test_freezesAtCompletion_id2() {
        let completed = date(origin).addingTimeInterval(100).ISO8601Format()
        let longAfter = date(origin).addingTimeInterval(5000)
        XCTAssertEqual(
            AmbientClock.display(firstFillAt: origin, completedAt: completed, now: longAfter),
            "1:40")
    }

    func test_hourRollover_growsADigitInsteadOfWrapping() {
        XCTAssertEqual(AmbientClock.display(seconds: 3599), "59:59")
        XCTAssertEqual(AmbientClock.display(seconds: 3600), "1:00:00")
        XCTAssertEqual(AmbientClock.display(seconds: 3725), "1:02:05")
    }

    func test_clockSkew_neverShowsNegativeTime() {
        let before = date(origin).addingTimeInterval(-30)
        XCTAssertEqual(
            AmbientClock.display(firstFillAt: origin, completedAt: nil, now: before),
            "0:00")
    }
}

File: apps/android/ui/src/test/kotlin/crossy/ui/AmbientClockTests.kt
// The shared ambient clock pinned against apps/ios AmbientClock.swift (ID-2; root DESIGN.md D15):
// 0:00 before the first fill, ticking from firstFillAt, frozen at the terminal instant, never
// negative under clock skew, and the m:ss / h:mm:ss display. Timestamps stay ISO 8601 wire strings;
// the parse tolerates the JS server's milliseconds and the fixture's plain form.
package crossy.ui

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test

class AmbientClockTests {
    @Test
    fun `ID2 parses wire timestamps with and without fractional seconds`() {
        assertEquals(0L, AmbientClock.parse("1970-01-01T00:00:00Z"))
        assertEquals(1_500L, AmbientClock.parse("1970-01-01T00:00:01.500Z"))
        // The explicit-offset form some tooling emits still parses.
        assertEquals(0L, AmbientClock.parse("1970-01-01T00:00:00+00:00"))
        assertNull(AmbientClock.parse("not a timestamp"))
    }

    @Test
    fun `ID2 reads zero before the first fill, quietly`() {
        assertEquals(0, AmbientClock.elapsedSeconds(firstFillAtMillis = null, freezeAtMillis = null, nowMillis = 99_000))
        assertEquals("0:00", AmbientClock.display(firstFillAt = null, freezeAt = null, nowMillis = 99_000))
    }

    @Test
    fun `ID2 ticks from the first fill while the room runs`() {
        assertEquals(65, AmbientClock.elapsedSeconds(firstFillAtMillis = 10_000, freezeAtMillis = null, nowMillis = 75_000))
        assertEquals("1:05", AmbientClock.display(65))
    }

    @Test
    fun `ID2 freezes at the terminal instant, now no longer moves it`() {
        // completedAt (or abandonedAt, the same freeze seam) pins the clock; a later now is ignored.
        assertEquals(30, AmbientClock.elapsedSeconds(firstFillAtMillis = 0, freezeAtMillis = 30_000, nowMillis = 999_000))
        assertEquals(
            "0:30",
            AmbientClock.display(
                firstFillAt = "1970-01-01T00:00:00Z",
                freezeAt = "1970-01-01T00:00:30Z",
                nowMillis = 999_000,
            ),
        )
    }

    @Test
    fun `ID2 clock skew can never show a negative time`() {
        assertEquals(0, AmbientClock.elapsedSeconds(firstFillAtMillis = 50_000, freezeAtMillis = null, nowMillis = 40_000))
        assertEquals("0:00", AmbientClock.display(-5))
    }

    @Test
    fun `DESIGN6 m colon ss under an hour, h colon mm colon ss from there`() {
        assertEquals("0:00", AmbientClock.display(0))
        assertEquals("0:09", AmbientClock.display(9))
        assertEquals("59:59", AmbientClock.display(3599))
        assertEquals("1:00:00", AmbientClock.display(3600))
        assertEquals("2:05:09", AmbientClock.display(2 * 3600 + 5 * 60 + 9))
    }
}
